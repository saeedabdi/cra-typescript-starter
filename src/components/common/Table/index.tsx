import cn from 'classnames';
import { CSSProperties, ReactElement, useState } from 'react';
import { useTranslation } from 'react-i18next';

import {
    CheckBoxWrapper,
    EmptyComponentsWrapper,
    LoadingComponentsWrapper,
    Table as TableStyle,
    TableCell,
    Th,
    Thead,
} from './table.style';

export interface TheadType<T> {
    title: string;
    key: string;
    style?: CSSProperties;
    className?: string;
    render?: (param?: string, rowData?: T) => ReactElement | string;
}
export interface TableState<T> {
    allSelected: boolean;
    selectedCount: number;
    selectedRows: T[];
}
export interface TableProps<T> {
    columns: TheadType<T>[];
    data: Array<any>;
    onSelectedRowsChange?: (data: TableState<T>) => void;
    selectableRows?: boolean;
    keyExtractor: (item: T) => string;
    className?: Record<string, string> | string;
    isLoading?: boolean;
    theadClassName?: string;
}

function Table<T = any>({
    data,
    className,
    keyExtractor,
    columns,
    onSelectedRowsChange,
    isLoading,
    theadClassName,
    selectableRows,
}: TableProps<T>): JSX.Element {
    const { t } = useTranslation();
    const [tableStates, setTableState] = useState<TableState<T>>({
        allSelected: false,
        selectedCount: 0,
        selectedRows: [],
    });
    const handleCheckBoxChange = (
        event: React.ChangeEvent<HTMLInputElement>,
        rowData: T | any,
        index: number,
    ) => {
        if (event.target.checked) {
            const selectedRows = [...tableStates.selectedRows, rowData];
            const newTableStates = {
                allSelected: data.length === selectedRows.length,
                selectedCount: selectedRows.length + 1,
                selectedRows: selectedRows,
            };
            setTableState(newTableStates);
            onSelectedRowsChange?.(newTableStates);
        } else {
            const newTableState = {
                allSelected: false,
                selectedCount: tableStates.selectedRows.length + 1,
                selectedRows: tableStates.selectedRows.filter(
                    (item) => keyExtractor(item) !== keyExtractor(rowData),
                ),
            };
            setTableState(newTableState);
            onSelectedRowsChange?.(newTableState);
        }
    };
    const handleCheckedAllChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        if (e.target.checked) {
            const newTableState = {
                allSelected: true,
                selectedCount: data.length + 1,
                selectedRows: data,
            };
            setTableState(newTableState);
            onSelectedRowsChange?.(newTableState);
        } else {
            const newTableState = {
                allSelected: false,
                selectedCount: 0,
                selectedRows: [],
            };
            onSelectedRowsChange?.(newTableState);
            setTableState(newTableState);
        }
    };

    return (
        <TableStyle className={cn(className)}>
            <thead className={theadClassName}>
                <tr>
                    {selectableRows && (
                        <Thead key="checkBox">
                            <div>
                                <input
                                    checked={tableStates.allSelected && data.length !== 0}
                                    onChange={handleCheckedAllChange}
                                    type="checkbox"
                                    name="select-all-rows"
                                />
                            </div>
                        </Thead>
                    )}
                    {columns.map((th) => (
                        <Th data-title={th.title} key={th.key}>
                            <div style={th.style}>
                                <span>{th.title}</span>
                            </div>
                        </Th>
                    ))}
                </tr>
            </thead>
            <tbody className="">
                {isLoading ? (
                    <tr>
                        <td colSpan={columns.length + 1}>
                            <LoadingComponentsWrapper className=" ">
                                <div>loading...</div>
                            </LoadingComponentsWrapper>
                        </td>
                    </tr>
                ) : data.length === 0 ? (
                    <tr>
                        <td colSpan={columns.length + 1}>
                            <EmptyComponentsWrapper className=" ">
                                {t('No data')}
                            </EmptyComponentsWrapper>
                        </td>
                    </tr>
                ) : (
                    data.map((tb, index) => {
                        return (
                            <tr
                                style={{
                                    height: 1,
                                }}
                                key={keyExtractor(tb)}
                            >
                                {selectableRows && (
                                    <CheckBoxWrapper>
                                        <label>
                                            <input
                                                type="checkbox"
                                                onChange={(e) => handleCheckBoxChange(e, tb, index)}
                                                checked={
                                                    !!tableStates.selectedRows.find(
                                                        (tableRow: T) =>
                                                            keyExtractor(tableRow) ===
                                                            keyExtractor(tb),
                                                    )
                                                }
                                                name={`select-row-${index}`}
                                            />
                                        </label>
                                    </CheckBoxWrapper>
                                )}
                                {columns.map((th) => {
                                    return (
                                        <TableCell
                                            style={{ height: '' }}
                                            data-title={th.title}
                                            key={th.key}
                                        >
                                            <div style={th.style}>
                                                <div className={cn(th.className)}>
                                                    {th.render
                                                        ? th.render(tb[th.key], tb)
                                                        : tb[th.key]}
                                                </div>
                                            </div>
                                        </TableCell>
                                    );
                                })}
                            </tr>
                        );
                    })
                )}
            </tbody>
        </TableStyle>
    );
}
export default Table;
