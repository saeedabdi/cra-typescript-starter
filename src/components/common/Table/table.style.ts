import styled from '@emotion/styled';
import { EmotionThemeInterface } from 'styles/theme/Theme';

export const Table = styled.table`
    position: relative;
    width: 100%;
`;
export const CheckBoxWrapper = styled.td`
    padding: 1rem 0rem 0.5rem 0rem;
    white-space: nowrap;
    label {
        width: 100%;
        display: flex;
        justify-content: center;
        input {
            width: 1.25rem;
            height: 1.25rem;
        }
    }
`;

export const Thead = styled.th`
    padding: 1rem 0rem 0.5rem 0rem;
    white-space: nowrap;
    div {
        padding: 0rem 0, 5rem;
        input {
            width: 1.25rem;
            height: 1.25rem;
        }
    }
`;
export const Th = styled.th`
    padding: 1rem 0rem 0.5rem 0rem;
    white-space: nowrap;
    div {
        color: ${({ theme }) => (theme as EmotionThemeInterface).colors?.accent8};
        padding: 0rem 0.5rem;
    }
`;
export const LoadingComponentsWrapper = styled.div`
    width: 100%;
    height: 16rem;
    color: ${({ theme }) => (theme as EmotionThemeInterface).colors?.accent8};
    display: flex;
    justify-content: center;
    align-items: center;
`;
export const EmptyComponentsWrapper = styled.div`
    width: 100%;
    height: 16rem;
    color: ${({ theme }) => (theme as EmotionThemeInterface).colors?.accent8};
    display: flex;
    justify-content: center;
    align-items: center;
`;
export const TableCell = styled.td`
    height: inherit;
    padding: 0;
    color: ${({ theme }) => (theme as EmotionThemeInterface).colors?.accent8};
    div {
        div {
            padding: 0.25rem 0rem;
            height: 100%;
            justify-content: center;
            align-items: center;
        }
    }
`;
