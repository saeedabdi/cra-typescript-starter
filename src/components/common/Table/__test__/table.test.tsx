import { fireEvent, render } from '@testing-library/react';
import i18n from 'i18n';
import renderer from 'react-test-renderer';

import Table, { TableState, TheadType } from '../index';

interface TestType {
    id: number;
    some: {
        name: string;
    };
}
type MockDataType = {
    columns: TheadType<TestType>[];
    data: {
        id: number;
        some: {
            name: string;
        };
    }[];
};

const dataMock = (): MockDataType => {
    return {
        columns: [
            {
                title: 'test',
                key: 'some',
                render: (param, row) => <>{row?.some?.name ?? ''}</>,
            },
        ],
        data: [
            {
                id: 1,
                some: { name: 'Apple' },
            },
            {
                id: 2,
                some: { name: 'Zuchinni' },
            },
        ],
    };
};

describe('Table component test', () => {
    beforeAll(() => {
        const i = i18n;
    });
    it('should render and empty table correctly', () => {
        expect(
            renderer
                .create(<Table keyExtractor={(item) => String(item.id)} columns={[]} data={[]} />)
                .toJSON(),
        ).toMatchSnapshot();
    });
    it('should render the correctly when using render function', () => {
        type TestData = {
            id: number;
            name: string;
        };

        const data: TestData[] = [{ id: 1, name: 'name' }];

        const tree = renderer
            .create(
                <Table<TestData>
                    data={data}
                    keyExtractor={(item) => String(item.id)}
                    columns={[
                        { title: 'name', key: 'name', render: (params, row) => <>{row?.name}</> },
                    ]}
                />,
            )
            .toJSON();

        expect(tree).toMatchSnapshot();
    });
});
describe('Table::onSelectedRowsChange', () => {
    it('should call onSelectedRowsChange with the correct values when select all rows is selected', () => {
        const mock = dataMock();
        const updatedMock = jest.fn();
        const { container } = render(
            <Table
                data={mock.data}
                keyExtractor={(item) => String(item.id)}
                columns={mock.columns}
                selectableRows
                onSelectedRowsChange={updatedMock}
            />,
        );
        fireEvent.click(
            container.querySelector('input[name="select-all-rows"]') as HTMLInputElement,
        );
        expect(updatedMock).toBeCalledWith({
            allSelected: true,
            selectedCount: mock.data.length + 1,
            selectedRows: mock.data,
        } as TableState<TestType>);
    });
    it('should call onSelectedRowsChange with the correct values when all rows are selected', () => {
        const mock = dataMock();
        const updatedMock = jest.fn();
        const { container } = render(
            <Table
                data={mock.data}
                columns={mock.columns}
                keyExtractor={(item) => String(item.id)}
                selectableRows
                onSelectedRowsChange={updatedMock}
            />,
        );
        fireEvent.click(container.querySelector('input[name="select-row-0"]') as HTMLInputElement);
        fireEvent.click(container.querySelector('input[name="select-row-1"]') as HTMLInputElement);
        expect(updatedMock).toBeCalledWith({
            allSelected: true,
            selectedCount: mock.data.length + 1,
            selectedRows: mock.data,
        } as TableState<TestType>);
    });
});
describe('data prop changes', () => {
    test('should update state if the data prop changes', () => {
        const mock = dataMock();
        const { container, rerender } = render(
            <Table
                keyExtractor={(item) => String(item.id)}
                data={mock.data}
                columns={mock.columns}
            />,
        );
        rerender(
            <Table
                keyExtractor={(item) => String(item.id)}
                data={[{ id: 1, some: { name: 'Someone else' } }]}
                columns={mock.columns}
            />,
        );
        expect(container.firstChild).toMatchSnapshot();
    });
});
