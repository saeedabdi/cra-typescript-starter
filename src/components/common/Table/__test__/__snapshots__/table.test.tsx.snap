// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component test should render and empty table correctly 1`] = `
<table
  className=" css-twyho5"
>
  <thead>
    <tr />
  </thead>
  <tbody
    className=""
  >
    <tr>
      <td
        colSpan={1}
      >
        <div
          className="  css-v9ada3"
        >
          No data
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table component test should render the correctly when using render function 1`] = `
<table
  className=" css-twyho5"
>
  <thead>
    <tr>
      <th
        className="css-phr1e7"
        data-title="name"
      >
        <div>
          <span>
            name
          </span>
        </div>
      </th>
    </tr>
  </thead>
  <tbody
    className=""
  >
    <tr
      style={
        Object {
          "height": 1,
        }
      }
    >
      <td
        className="css-1fp9v71"
        data-title="name"
        style={
          Object {
            "height": "",
          }
        }
      >
        <div>
          <div
            className=""
          >
            name
          </div>
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`data prop changes should update state if the data prop changes 1`] = `
<table
  class=" css-twyho5"
>
  <thead>
    <tr>
      <th
        class="css-phr1e7"
        data-title="test"
      >
        <div>
          <span>
            test
          </span>
        </div>
      </th>
    </tr>
  </thead>
  <tbody
    class=""
  >
    <tr
      style="height: 1px;"
    >
      <td
        class="css-1fp9v71"
        data-title="test"
      >
        <div>
          <div
            class=""
          >
            Someone else
          </div>
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;
